File: college-tycoon/build-standalone.js
#!/usr/bin/env node
/* Inline css/ and js/ into one self-contained HTML file.
   Used to publish the game somewhere that can only serve a single page.

     node college-tycoon/build-standalone.js [outfile]

   Reads the real sources, so the bundle can never drift from the repo. */

const fs = require("fs");
const path = require("path");

/* --fragment drops the document skeleton, for hosts that supply their own
   <!doctype>/<head>/<body> and only want the page content. */
const args = process.argv.slice(2);
const fragment = args.includes("--fragment");

const root = __dirname;
const out = args.find((a) => !a.startsWith("--")) || path.join(root, "college-tycoon.html");

const read = (p) => fs.readFileSync(path.join(root, p), "utf8");

/* Script order matters: util defines helpers the data tables call at runtime. */
const SCRIPTS = ["js/util.js", "js/data.js", "js/engine.js", "js/ui.js", "js/main.js"];

let html = read("index.html");

html = html.replace(
  /<link rel="stylesheet" href="css\/game\.css">/,
  `<style>\n${read("css/game.css")}\n</style>`);

for (const src of SCRIPTS) {
  const tag = `<script src="${src}"></script>`;
  if (!html.includes(tag)) throw new Error(`index.html no longer loads ${src}`);
  html = html.replace(tag, `<script>\n${read(src)}\n</script>`);
}

/* data: URIs are self-contained; anything else would be a network fetch. */
const external = html.match(/<(?:link|script|img)[^>]+(?:href|src)="(?!data:)([^"]*)"/g);
if (external) {
  throw new Error("bundle still references external files:\n  " + external.join("\n  "));
}

if (fragment) {
  const title = html.match(/<title>[\s\S]*?<\/title>/)[0];
  const style = html.match(/<style>[\s\S]*?<\/style>/)[0];
  const body = html.match(/<body>([\s\S]*)<\/body>/)[1];
  html = `${title}\n${style}\n${body.trim()}\n`;
}

fs.writeFileSync(out, html);
console.log(`${out} — ${(Buffer.byteLength(html) / 1024).toFixed(0)} KB${fragment ? " (fragment)" : ""}`);

File: tools/build-standalone.py
#!/usr/bin/env python3
"""Inline Git Quest into one self-contained HTML file.

    python3 tools/build-standalone.py [out.html] [--keep-report] [--artifact]

The result needs no server and no network: styles and scripts are inlined, so
it can be e-mailed to students, dropped on a USB stick, or published as a
single page. The "report to lecturer" panel is dropped by default because it
POSTs to Google Apps Script, which sandboxed hosts block; pass --keep-report
to keep it for a normal web host.

--artifact emits the same page as a fragment (no <html>/<head>/<body> wrapper)
for hosts that supply their own document skeleton, such as Claude Artifacts.
"""
import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
SCRIPTS = ["js/git-engine.js", "js/git-viz.js", "js/git-levels.js", "js/git-portal.js"]


def main() -> int:
    args = [a for a in sys.argv[1:] if not a.startswith("--")]
    keep_report = "--keep-report" in sys.argv
    artifact = "--artifact" in sys.argv
    out = Path(args[0]) if args else ROOT / "git-quest-standalone.html"

    page = (ROOT / "git-game.html").read_text(encoding="utf-8")
    body = re.search(r"<body[^>]*>(.*)</body>", page, re.S).group(1)
    body = re.sub(r'\s*<script src="[^"]+"></script>', "", body)

    if not keep_report:
        body = re.sub(r'\s*<section class="gq-panel gq-report">.*?</section>', "", body, flags=re.S)

    css = (ROOT / "css/git-game.css").read_text(encoding="utf-8")
    js = "\n\n".join((ROOT / s).read_text(encoding="utf-8") for s in SCRIPTS)

    if artifact:
        # The host writes <body> for us, so claim the theme from script, and
        # point the footer at the repo instead of the local assessment page.
        body = body.replace(
            '<a href="index.html">back to the C++ assessment</a>',
            '<a href="https://github.com/SCLim999/pps2114-quiz">source on GitHub</a>')
        out.write_text(
            "<title>Git Quest</title>\n"
            '<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>\n'
            '<link rel="stylesheet" href="https://fonts.googleapis.com/css2'
            '?family=JetBrains+Mono:wght@400;700&display=swap">\n'
            "<style>\n" + css + "\n</style>\n" + body.strip() +
            "\n<script>\ndocument.body.classList.add('gq');\n" + js + "\n</script>\n",
            encoding="utf-8")
        print(f"wrote {out} (artifact fragment)")
        return 0

    html = (
        "<!DOCTYPE html>\n<html lang=\"en\">\n<head>\n<meta charset=\"UTF-8\">\n"
        '<meta name="viewport" content="width=device-width, initial-scale=1.0">\n'
        "<title>Git Quest</title>\n<style>\n" + css + "\n</style>\n</head>\n"
        '<body class="gq">\n' + body.strip() + "\n<script>\n" + js + "\n</script>\n</body>\n</html>\n"
    )
    out.write_text(html, encoding="utf-8")
    print(f"wrote {out} ({len(html) / 1024:.0f} KB)")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
